db.go

```diff
@@ -127,48 +154,119 @@ func (x *dummyAuthenticator) ResetPasswordFinish(identity string, token string,
 		return nil
 	} else {
 		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) RenameIdentity(oldIdent, newIdent string) error {
+func (x *dummyUserStoreAndAuth) UpdateIdentity(userId UserId, email, username, firstname, lastname, mobilenumber string) error {
 	x.usersLock.Lock()
 	defer x.usersLock.Unlock()
+	if user, exists := x.users[userId]; exists {
+		user.email = email
+		user.username = username
+		user.firstname = firstname
+		user.lastname = lastname
+		user.mobilenumber = mobilenumber
+	} else {
+		return ErrIdentityAuthNotFound
+	}
+	return nil
+}
 
-	if _, exists := x.users[CanonicalizeIdentity(newIdent)]; exists {
-		return ErrIdentityExists
+func (x *dummyUserStoreAndAuth) ArchiveIdentity(userId UserId) error {
+	x.usersLock.Lock()
+	defer x.usersLock.Unlock()
+	if user, exists := x.users[userId]; exists {
+		user.archived = true
+	} else {
+		return ErrIdentityAuthNotFound
 	}
+	return nil
+}
 
-	oldIdent = CanonicalizeIdentity(oldIdent)
-	newKey := CanonicalizeIdentity(newIdent)
+func (x *dummyUserStoreAndAuth) RenameIdentity(oldEmail, newEmail string) error {
+	x.usersLock.Lock()
+	defer x.usersLock.Unlock()
 
-	if user, exists := x.users[oldIdent]; exists {
-		delete(x.users, oldIdent)
-		x.users[newKey] = &dummyUser{newIdent, user.password, ""}
-		return nil
+	newKey := CanonicalizeIdentity(newEmail)
+	oldEmail = CanonicalizeIdentity(oldEmail)
+	newUser := getUser(x.users, newKey)
+	if newUser == nil {
+		oldUser := getUser(x.users, oldEmail)
+
+		if oldUser != nil {
+			x.users[oldUser.userId].email = newEmail
+			return nil
+		} else {
+			return ErrIdentityAuthNotFound
+		}
 	} else {
-		return ErrIdentityAuthNotFound
+		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) GetIdentities() ([]string, error) {
+func (x *dummyUserStoreAndAuth) GetIdentities() ([]AuthUser, error) {
 	x.usersLock.RLock()
-	list := []string{}
-	for identity, _ := range x.users {
-		list = append(list, identity)
+	defer x.usersLock.RUnlock()
+
+	list := []AuthUser{}
+	for _, v := range x.users {
+		if v.archived == false {
+			list = append(list, AuthUser{v.userId, v.email, v.username, v.firstname, v.lastname, v.mobilenumber})
+		}
 	}
-	x.usersLock.RUnlock()
 	return list, nil
 }
 
-func (x *dummyAuthenticator) Close() {
+func (x *dummyUserStoreAndAuth) GetIdentityFromUserId(userId UserId) (string, error) {
+	x.usersLock.RLock()
+	defer x.usersLock.RUnlock()
+
+	for _, v := range x.users {
+		if v.userId == userId {
+			return v.email, nil
+		}
+	}
+
+	return "", ErrIdentityAuthNotFound
+}
+
+func (x *dummyUserStoreAndAuth) GetUserIdFromIdentity(identity string) (UserId, error) {
+	x.usersLock.RLock()
+	defer x.usersLock.RUnlock()
+
+	if CanonicalizeIdentity(identity) == "" {
+		return -1, ErrIdentityEmpty
+	}
+
+	for _, v := range x.users {
+		if v.email == identity {
+			return v.userId, nil
+		}
+	}
+	return -1, ErrIdentityAuthNotFound
+}
+
+func getUser(users map[UserId]*dummyUser, email string) *dummyUser {
```

Comment:
Since `getUser` function only ever operates on members of dummyUserStoreAndAuth, I would make it a member of dummyUserStoreAndAuth. This is just to narrow the scope of things.
Same goes for `generateUserId`.